core/src/main/java/net/opentsdb/query/processor/downsample/DownsampleNumericToNumericArrayIterator.java

```diff
@@ -0,0 +1,577 @@
+    }
+    if (node.config() == null) {
+      throw new IllegalArgumentException("Node config cannot be null.");
+    }
+    this.result = (DownsampleResult) result;
+    this.node = node;
+
+    config = (DownsampleConfig) node.config();
+
+    this.aggEnum = AggEnum.valueOf(config.getAggregator().toLowerCase());
+    interval_start = this.result.start().getCopy();
+    interval_end = this.result.end().getCopy();
+    intervals = ((DownsampleConfig) node.config()).intervals();
+    intervalPart = (int) ((DownsampleConfig) node.config()).interval().get(ChronoUnit.SECONDS);
+
+    final Optional<TypedTimeSeriesIterator<? extends TimeSeriesDataType>> optional =
+        source.iterator(NumericType.TYPE);
+    if (optional.isPresent()) {
+      iterator = optional.get();
+    } else {
+      iterator = null;
+    }
+
+    has_next = iterator != null ? iterator.hasNext() : false;
+
+  }
+
+  @Override
+  public boolean hasNext() {
+    return has_next;
+  }
+
+  /**
+   * The pre computed result in array aggregator
+   * 
+   * @param aggregator
+   * @return
+   */
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  @Override
+  public TimeSeriesValue<NumericArrayType> nextPool(Aggregator aggregator) {
+
+    has_next = false;
+
+    NumericArrayAggregator numericArrayAggregator = (NumericArrayAggregator) aggregator; // group by
+                                                                                         // aggregator
+    Accumulator accumulator = DownsampleNumericToNumericArrayIterator.threadLocalAccs.get();
+    accumulator.reset();
+
+    double[] localDoubleAggs = DownsampleNumericToNumericArrayIterator.threadLocalDoubleAggs.get();
+    long[] localLongAggs = DownsampleNumericToNumericArrayIterator.threadLocalLongAggs.get();
+
+    resetLocalAggs(localDoubleAggs, localLongAggs);
+
+    TimeStamp nextTs = null;
+
+    boolean flush = false;
+    int index = 0;
+    while (iterator.hasNext()) {
+      TimeSeriesValue<NumericType> next = (TimeSeriesValue<NumericType>) iterator.next();
+
+      if (next.timestamp().compare(Op.GTE, this.interval_start)
+          && next.timestamp().compare(Op.LT, this.interval_end)) {
+
+        if (nextTs == null) {
+          // Use the first valid dp timestamp as the first Timestamp.
+          nextTs = next.timestamp().getCopy();
+          long nextRoundoffTs =
+              next.timestamp().epoch() - (next.timestamp().epoch() % intervalPart) + intervalPart;
+          nextTs.updateEpoch(nextRoundoffTs);
+        }
+
+        if (next.timestamp().compare(Op.GTE, nextTs)) {
+          // flush previous interval once new interval starts
+          flush = true;
+
+          // find the index to update in the Aggregator
+          index = (int) (((nextTs.epoch() - interval_start.epoch() - intervalPart) / intervalPart)
+              % intervals);
+
+          // get the next timestamp in the downsample
+          long nextTsRoundoff =
```

Comment:
And there's a `timestamp.add(config.interval())` to use here that'll save some cycles.

Reply:
We cant use config.interval since the storage iterator would only give the next non null interval datapoint that might not be the immediate next dp.